Cabal/tests/PackageTests/PackageTester.hs

```diff
@@ -1,367 +1,540 @@
+data Result = Result
+    { resultExitCode :: ExitCode
+    , resultDirectory :: Maybe FilePath
+    , resultCommand :: String
+    , resultOutput :: String
+    } deriving Show
 
-assertRegisterSucceeded :: Result -> Assertion
-assertRegisterSucceeded result = unless (successful result) $
-    assertFailure $
-    "expected: \'setup register\' should succeed\n" ++
-    "  output: " ++ outputText result
+requireSuccess :: Result -> TestM Result
+requireSuccess r@Result { resultCommand = cmd
+                        , resultExitCode = exitCode
+                        , resultOutput = output } = do
+    (_, test) <- ask
+    let lastN n = unlines . reverse . take n . reverse . lines
+    when (exitCode /= ExitSuccess && not (testShouldFail test)) $
+        assertFailure $ "Command " ++ cmd ++ " failed.\n" ++
+        "last 5 lines of output (see test.log for full):\n" ++ lastN 5 output
+    when (exitCode == ExitSuccess && testShouldFail test) $
+        assertFailure $ "Command " ++ cmd ++ " succeeded.\n" ++
+        "last 5 lines of output (see test.log for full):\n" ++ lastN 5 output
+    -- TODO: breadcrumb for test.log for full output
+    return r
 
-assertRegisterFailed :: Result -> Assertion
-assertRegisterFailed result = when (successful result) $
-    assertFailure $
-    "expected: \'setup register\' should fail\n" ++
-    "  output: " ++ outputText result
+record :: Result -> TestM ()
+record res = do
+    build_dir <- distDir
+    liftIO $ createDirectoryIfMissing True build_dir
+    -- TODO: make a runnable shell script
+    liftIO $ C.appendFile (build_dir </> "test.log")
+                         (C.pack $ "+ " ++ resultCommand res ++ "\n"
+                            ++ resultOutput res)
 
+------------------------------------------------------------------------
+-- * Test helpers
 
-assertOutputContains :: String -> Result -> Assertion
+assertFailure :: MonadIO m => String -> m ()
+assertFailure = liftIO . HUnit.assertFailure
+
+assertEqual :: (Eq a, Show a, MonadIO m) => String -> a -> a -> m ()
+assertEqual s x y = liftIO $ HUnit.assertEqual s x y
+
+assertBool :: MonadIO m => String -> Bool -> m ()
+assertBool s x = liftIO $ HUnit.assertBool s x
+
+shouldExist :: MonadIO m => FilePath -> m ()
+shouldExist path = liftIO $ doesFileExist path >>= assertBool (path ++ " should exist")
+
+shouldNotExist :: MonadIO m => FilePath -> m ()
+shouldNotExist path =
+    liftIO $ doesFileExist path >>= assertBool (path ++ " should exist") . not
+
+shouldFail :: TestM a -> TestM a
+shouldFail = withReaderT (\(suite, test) -> (suite, test { testShouldFail = not (testShouldFail test) }))
+
+whenGhcVersion :: (Version -> Bool) -> TestM () -> TestM ()
+whenGhcVersion p m = do
+    (suite, _) <- ask
+    when (p (ghcVersion suite)) m
+
+withPackage :: FilePath -> TestM a -> TestM a
+withPackage f = withReaderT (\(suite, test) -> (suite, test { testCurrentPackage = f }))
+
+-- TODO: Really should accumulate... but I think to do this
```

Comment:
I think this should be added before merge, or the comment should be removed. Doing this post-merge would change the semantics of withEnv rather subtly -- now it might not matter greatly, but then again it might.